src/proxy/connection_manager.rs

```diff
@@ -0,0 +1,359 @@
+    #[inline(always)]
+    async fn drain(self) {
+        drop(self.rx); // very important, drain cannont complete if there are outstand rx
+        self.tx.drain().await;
+    }
+}
+
+#[derive(Clone)]
+pub struct ConnectionManager {
+    drains: Arc<RwLock<HashMap<Connection, ConnectionDrain>>>,
+}
+
+impl ConnectionManager {
+    pub fn new() -> Self {
+        ConnectionManager {
+            drains: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    // register a connection with the manager and get a channel to receive on
+    pub async fn track(self, c: &Connection) -> drain::Watch {
+        let cd = ConnectionDrain::new();
+        let rx = cd.rx.clone();
+        let mut drains = self.drains.write().await;
+        if let Some(w) = drains.remove(c) {
+            w.count.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+            let rx = w.rx.clone();
+            drains.insert(c.clone(), w);
+            debug!("there are {} tracked connections", drains.len());
+            return rx;
+        }
+        drains.insert(c.clone(), cd);
+        debug!("there are {} tracked connections", drains.len());
+        rx
+    }
+
+    // releases tracking on a connection
+    // uses a counter to determine if there are other tracked connections or not so it may retain the tx/rx channels when necessary
+    pub async fn release(self, c: &Connection) {
+        let mut drains = self.drains.write().await;
+        if let Some((k, v)) = drains.remove_entry(c) {
+            if v.count.fetch_sub(1, std::sync::atomic::Ordering::SeqCst) > 1 {
+                // something else is tracking this connection, retain
+                drains.insert(k, v);
+            }
+        }
+    }
+
+    // signal all connections listening to this channel to take action (typically terminate traffic)
+    async fn close(&self, c: &Connection) {
+        if let Some(cd) = self.drains.clone().write().await.remove(c) {
+            cd.drain().await;
+        } else {
+            // this is bad, possibly drain called twice
+            error!("requested drain on a Connection which wasn't initialized");
+        }
+    }
+
+    //  get a list of all connections being tracked
+    async fn connections(&self) -> Vec<Connection> {
+        // potentially large copy under read lock, could require optomization
+        self.drains.read().await.keys().cloned().collect()
+    }
+}
+
+pub async fn policy_watcher(
+    state: DemandProxyState,
+    mut stop_rx: watch::Receiver<()>,
+    connection_manager: ConnectionManager,
+    parent_proxy: &str,
+) {
+    let mut policies_changed = state.read().policies.subscribe();
+    loop {
+        tokio::select! {
+            _ = stop_rx.changed() => {
+                break;
+            }
+            _ = policies_changed.changed() => {
+                let connections = connection_manager.connections().await;
+                for conn in connections {
+                    if !state.assert_rbac(&conn).await {
```

Comment:
Sorry, I'm not a Rust expert - does this await individually or globally? Any time there's IO, you probably don't want to stall in a loop in a switch - you may overflow the pending events because IO is slow. It's best to fire-and-forget and then let someone reap the zombies.

Reply:
This will await the individual assertion against the one connection but await doesn't mean block (there is a specific block function if that's what you want though). If the Future being awaited is not ready it relinquishes control and will be polled later by the executor of the task/thread.

Rusts Futures are somewhat unique in that they are lazy. If nothing drives them to completion they will not do any work so we can't totally fire and forget. We can either await them and let the executor/callback model poll them or we could collect the all futures and join them later. If we do neither no work would happen though.

Reply:
I feel like you do want to block the loop or at least debounce it to avoid edge-trigerred event spam in the runtime. But I have no idea how to express that in Rust. Maybe a separate actor that is "fast" in consuming policy events but "slow" in enacting them.

Reply:
The code you are commenting on is _sort of_ that actor (in concept at least). It's being spun up in its own task so it should be (at least somewhat) independent of the tasks which are actually moving data.

Reply:
If I understand the event spam, is the concern is we enqueue multiple times to trigger this code? Like while we are looping here, we get `changed()` multiple times stacked up?

If so, I think it will have at most one pending. That being said, you could still hit some spam I guess

Reply:
I've changed the implementation of sending to policy subscribers to be less frequent in cases where we've got batched updates. Should help